### PR TITLE
18ka pullman

### PR DESCRIPTION
**Pull request by ventusignis**
image
image
image
image
image

### CODE REVIEW

File: lib/engine/game/g_18_ka/game.rb [hunk 1]
```diff
@@ -923,6 +942,69 @@ def setup
           national.add_ability(self.class::NATIONAL_FORCED_WITHHOLD_ABILITY)
         end
 
+        def route_distance_str(route)
+          towns = route.visited_stops.count(&:town?)
+          cities = route_distance(route) - towns
+          "#{cities}+#{towns}"
+        end
+
+        def revenue_for(route, stops)
```
Comment (tobymao) [inline lib/engine/game/g_18_ka/game.rb:951]: why do you need this?

File: lib/engine/game/g_18_ka/step/route.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                  },
+                ]
+              end
+            end
+            @pullman_train_assignments = {}
+          end
+
+          def train_choices(entity)
+            # Diesels don't get pullmans. That'd be silly.
+            Array(@game.route_trains(entity).reject { |t| @game.pullman_train?(t) || t.variant['name'] == 'D' })
```
Comment (tobymao) [inline lib/engine/game/g_18_ka/step/route.rb:81]: why are you wrapping this in array?

File: lib/engine/game/g_18_ka/step/route.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          end
+
+          def pullman_choices(entity)
+            Array(entity.trains.select { |t| @game.pullman_train?(t) && !@pullman_train_assignments[t] })
```
Comment (tobymao) [inline lib/engine/game/g_18_ka/step/route.rb:85]: same

File: lib/engine/game/g_18_ka/step/route.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          def process_run_routes(action)
+            super
+
+            detach_pullmans unless @pullman_train_assignments.empty?
+          end
+
+          def train_city_distance(train)
+            return train.distance if train.distance.is_a?(Numeric)
+
+            distance_city = train.distance.find { |n| n['nodes'].length > 1 }
```
Comment (tobymao) [inline lib/engine/game/g_18_ka/step/route.rb:106]: size

File: lib/engine/game/g_18_ka/step/route.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            distance_city ? distance_city['visit'] : 0
+          end
+
+          def train_town_distance(train)
+            return 0 if train.distance.is_a?(Numeric)
+
```
Comment (tobymao) [inline lib/engine/game/g_18_ka/step/route.rb:112]: size

**Comment by tobymao** [comment]
does autorouting work?

File: lib/engine/game/g_18_ka/game.rb (hunk 1 above)
Comment (ventusignis) [reply lib/engine/game/g_18_ka/game.rb:951]: This is to reset the behavior of 1856's revenue_for, I didn't see an elegant way to call the base revenue_for so I re implemented it here

File: lib/engine/game/g_18_ka/step/route.rb (hunk 2 above)
Comment (ventusignis) [reply lib/engine/game/g_18_ka/step/route.rb:81]: I had `pullman_choices` using `find` originally, Array is no longer needed

File: lib/engine/game/g_18_ka/step/route.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/game/g_18_ka/step/route.rb:85]: changed

File: lib/engine/game/g_18_ka/step/route.rb (hunk 4 above)
Comment (ventusignis) [reply lib/engine/game/g_18_ka/step/route.rb:106]: changed

File: lib/engine/game/g_18_ka/step/route.rb (hunk 5 above)
Comment (ventusignis) [reply lib/engine/game/g_18_ka/step/route.rb:112]: changed

**Comment by ventusignis** [comment]
I was making an example and autorouting generated this invalid (can't duplicate track) route, not really sure why
image
I'm unable to duplicate the route by manually selecting locations

**Review by tobymao — APPROVED** [review]

**Pull request merged**